session/schema_amender.go

```diff
@@ -0,0 +1,563 @@
+}
+
+func (a *amendOperationAddNewIndex) genMutations(ctx context.Context, sctx sessionctx.Context, commitMutations tikv.CommitterMutations,
+	kvMap *rowKvMap) (tikv.CommitterMutations, error) {
+	resAddMutations := tikv.NewCommiterMutations(8)
+	for i, key := range commitMutations.GetKeys() {
+		keyOp := commitMutations.GetOps()[i]
+		if tablecodec.IsIndexKey(key) || tablecodec.DecodeTableID(key) != a.info.tblInfoAtCommit.Meta().ID {
+			continue
+		}
+		if !isInsertOp(keyOp) {
+			continue
+		}
+		addMutations, err := a.processKey(ctx, sctx, key, kvMap.newRowKvMap)
+		if err != nil {
+			return resAddMutations, err
+		}
+		if len(addMutations.GetKeys()) > 0 {
+			resAddMutations.MergeMutations(addMutations)
+		}
+	}
+	return resAddMutations, nil
+}
+
+func (a *amendOperationAddIndexInfo) genIndexKeyValue(ctx context.Context, sctx sessionctx.Context, kvMap map[string][]byte,
+	key []byte, kvHandle kv.Handle, keyOnly bool) ([]byte, []byte, error) {
+	colMap := make(map[int64]*types.FieldType)
+	for _, oldCol := range a.tblInfoAtStart.Meta().Cols() {
+		colMap[oldCol.ID] = &oldCol.FieldType
+	}
+	chk := a.chk
+	chk.Reset()
+	// The Op_Put may not exist in old value kv map.
+	val, ok := kvMap[string(key)]
+	if !ok && keyOnly {
+		return nil, nil, nil
+	}
+	err := executor.DecodeRowValToChunk(sctx, a.schemaAndDecoder.schema, a.tblInfoAtStart.Meta(), kvHandle, val, chk, a.schemaAndDecoder.decoder)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend decode value to chunk failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	idxVals := make([]types.Datum, 0, len(a.indexInfoAtCommit.Meta().Columns))
+	for _, oldCol := range a.relatedOldIdxCols {
+		idxVals = append(idxVals, chk.GetRow(0).GetDatum(oldCol.Offset, &oldCol.FieldType))
+	}
+
+	// Generate index key buf.
+	newIdxKey, distinct, err := tablecodec.GenIndexKey(sctx.GetSessionVars().StmtCtx,
+		a.tblInfoAtCommit.Meta(), a.indexInfoAtCommit.Meta(), a.tblInfoAtCommit.Meta().ID, idxVals, kvHandle, nil)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend generate index key failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	if keyOnly {
+		return newIdxKey, []byte{}, nil
+	}
+
+	// Generate index value buf.
+	var containsNonBinaryString bool
+	for _, idxCol := range a.indexInfoAtCommit.Meta().Columns {
+		col := a.tblInfoAtCommit.Meta().Columns[idxCol.Offset]
+		if col.EvalType() == types.ETString && !mysql.HasBinaryFlag(col.Flag) {
+			containsNonBinaryString = true
+			break
+		}
+	}
+	newIdxVal, err := tablecodec.GenIndexValue(sctx.GetSessionVars().StmtCtx, a.tblInfoAtCommit.Meta(),
+		a.indexInfoAtCommit.Meta(), containsNonBinaryString, distinct, false, idxVals, kvHandle)
+	if err != nil {
+		logutil.Logger(ctx).Warn("amend generate index values failed", zap.Error(err))
+		return nil, nil, errors.Trace(err)
+	}
+	return newIdxKey, newIdxVal, nil
+}
+
+func (a *amendOperationAddNewIndex) processKey(ctx context.Context, sctx sessionctx.Context, key []byte,
```

Comment:
The code would be simpler if we pass pointer of `resAddMutations` as an argument.

Reply:
Name it `processRowKey` is more readable.